org-code-javabuilder/media/src/main/java/org/code/media/AudioWriter.java

```diff
@@ -0,0 +1,65 @@
+package org.code.media;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+/**
+ * Writer for concatenating audio data from multiple audio sources. The raw audio samples in bytes
+ * are written to a raw output stream and finally written to the given audio output stream when
+ * complete.
+ */
+public class AudioWriter {
+  private final ByteArrayOutputStream audioOutputStream;
+  private final ByteArrayOutputStream rawOutputStream;
+
+  public AudioWriter(ByteArrayOutputStream audioOutputStream) {
+    this(audioOutputStream, new ByteArrayOutputStream());
+  }
+
+  AudioWriter(ByteArrayOutputStream audioOutputStream, ByteArrayOutputStream rawOutputStream) {
+    this.audioOutputStream = audioOutputStream;
+    this.rawOutputStream = rawOutputStream;
+  }
+
+  public void writeAudioSamples(double[] samples) throws SoundException {
+    final byte[] bytes = AudioUtils.convertDoubleArrayToByteArray(samples);
+    try {
+      this.rawOutputStream.write(bytes);
+    } catch (IOException e) {
+      throw new SoundException(e);
+    }
+  }
+
+  public void writeAudioFile(String filename) throws SoundException {
+    writeAudioSamples(SoundLoader.read(filename));
+  }
+
+  public void addDelay(int delayMs) throws SoundException {
+    final double[] emptySamples =
+        new double[(int) (((double) delayMs / 1000.0) * AudioUtils.getDefaultSampleRate())];
+    writeAudioSamples(emptySamples);
+  }
+
+  /**
+   * Writes the raw audio data in rawOutputStream to audioOutputStream in a valid audio file format
+   * and closes output streams.
+   *
+   * @throws SoundException
+   */
+  public void writeToAudioStreamAndClose() throws SoundException {
+    final byte[] bytes = this.rawOutputStream.toByteArray();
+    AudioUtils.writeBytesToOutputStream(bytes, this.audioOutputStream);
+
+    try {
+      this.rawOutputStream.close();
+      this.audioOutputStream.close();
```

Comment:
for my information: is the reason we need two streams that we need to write the `audioOutputStream` all at once?

Reply:
Yep, so `rawOutputStream` just collects the raw audio samples themselves, but when writing to a file we also need the appropriate file headers, etc which `AudioSystem.write()` provides for us. So `audioOutputStream` is used to write out the full contents of the audio file.

I think we could use one output stream if we collect the raw bytes, reset the stream, and write the full file back to that stream? Although I think eventually this will write out to a file rather than an output stream so this may go away.